compiler/luci/service/src/Nodes/CircleBatchMatMul.cpp

```diff
@@ -30,4 +52,56 @@ luci::CircleNode *CloneNodeLet<CN::ABC>::visit(const luci::CircleBatchMatMul *no
   return cloned;
 }
 
+// BatchMatMulV2 supports broadcasting in the batch dimensions(BatchMatMul doesn't)
+// TODO Distinguish BatchMatMul and BatchMatMulV2
+loco::TensorShape sinf::Algorithm::visit(const luci::CircleBatchMatMul *node)
+{
+  const auto x = loco::must_cast<CircleNode *>(node->x());
+  const auto y = loco::must_cast<CircleNode *>(node->y());
+
+  const auto x_shape = sinf::circle_shape(x);
+  const auto y_shape = sinf::circle_shape(y);
+
+  uint32_t x_rank = x_shape.rank();
+  uint32_t y_rank = y_shape.rank();
+  assert(x_rank >= 2 && y_rank >= 2);
```

Comment:
just to be sure, `x_rank` and `y_rank` must be `>= 2`
and if not this model is invalid ?

Reply:
> and if not this model is invalid ? 

Yes. I think we can consider 'input with `rank < 2` BatchMatMul is invalid'.  

Because both TFLite and Pytorch don't allow `rank < 2` inputs for BatchMatMul :  

* TFLite `ops::BatchMatMul` (https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/batch-mat-mul) allows 2-D or higher inputs 
  > Args: 
   > x: 2-D or higher with shape [..., r_x, c_x]. 
   > y: 2-D or higher with shape [..., r_y, c_y]. 

* pytorch `torch.bmm` (https://pytorch.org/docs/stable/generated/torch.bmm.html) allows only D inputs (>= 2D) 
  > input and mat2 must be 3-D tensors each containing the same number of matrices.

Reply:
Also, original shape inference uses assertion to check `rank >=2`. 
 https://github.com/Samsung/ONE/pull/13759/files#r1732652062

Reply:
I leave an additional comment about this - https://github.com/Samsung/ONE/pull/13759#discussion_r1732649791. 

I thought that the circle model is imported from TensorFlow or PyTorch, and if both frameworks do not allow inputs under 2D, such a circle model would not exist. So I thought we don't need to consider such cases. 

------------------------------------------

By the way, using an exception is also fine with me. If others think using an exception is better for safety, I'll fix it.

Reply:
+1 for exception. TF, Torch or our converting tools (circle-exir, onnx2circle) can have a bug. It would be better to explicitly check invalid inputs.